lib/SIL/Utils/Dominance.cpp

```diff
@@ -141,3 +141,131 @@ void PostDominanceInfo::verify() const {
+    callback(insts.front(), {});
+    return;
+  }
+
+  // Okay, handle the general case.
+
+  // This is essentially a sort of union-find.  We track the set of
+  // "dominance roots": mutually dominance-unrelated instructions, along
+  // with possibly-empty sets of instructions that each dominates.
+  // (If A and B are dominance-unrelated, and B dominates C, then A
+  // and C must be dominance-unrelated by the properties of dominance,
+  // so this is well-defined.)  When we discover new relationships, we
+  // may need to combine multiple existing sets, which we do lazily by
+  // moving all but one of the sets into the stale list and setting
+  // their MergeInfo links.  At the end, we'll collapse all these
+  // relationships.
+  struct DominanceRoot {
+    SILInstruction *Dominator;
+    DominanceRoot *TrueRoot = nullptr;
+    SmallVector<SILInstruction*, 7> Dominated;
+    DominanceRoot(SILInstruction *dominator) : Dominator(dominator) {}
+
+    bool isStale() const {
+      return TrueRoot != nullptr;
+    }
+
+    DominanceRoot *getTrueRoot() {
+      assert(isStale());
+      if (!TrueRoot->isStale()) return TrueRoot;
+      // Union-find: update our stored true root after making this call
+      // to ensure that computing true roots for all the stale roots
+      // is amortized linear.
+      TrueRoot = TrueRoot->getTrueRoot();
+      return TrueRoot;
+    }
+  };
+  using DominanceRootPtr = std::unique_ptr<DominanceRoot>;
+  SmallVector<DominanceRootPtr, 8> dominanceRoots;
+  SmallVector<DominanceRootPtr, 8> staleDominanceRoots;
+
+  // Iteratively compare each instruction with the current set of
+  // dominance roots.
+  for (SILInstruction *cur : insts) {
+    bool foundDominator = false;
+    DominanceRoot *foundDominatedRoot = nullptr;
+    for (unsigned ri = 0, re = dominanceRoots.size(); ri != re; ) {
+      auto &root = dominanceRoots[ri];
+      assert(!root->isStale());
+
+      // If the current instruction is dominated by the root (likely
+      // if the instruction array is in function order), add it to the
+      // root's dominated set.  But we don't need to check this if we've
+      // already found a dominated root because that would imply a
+      // dominance relationship between two existing roots, which should
+      // be impossible.
+      if (!foundDominatedRoot &&
+          dominance.properlyDominates(root->Dominator, cur)) {
+        root->Dominated.push_back(cur);
+
+        // We don't need to consider any other dominance roots: since
+        // roots have no dominance relationship with each other, and
+        // we are dominated by one of them, we cannot have a dominance
+        // relationship with any of the others.
+        foundDominator = true;
+        break;
+      }
+
+      // If the current instruction dominates the root, we need to remove
+      // or replace the other instruction in the dominance roots.
+      if (dominance.properlyDominates(cur, root->Dominator)) {
```

Comment:
``properlyDominates`` iterates over instructions in a block. This can result in quadratic complexity, e.g. in case of many ``hop_to_execute``s in a large block.